### ISSUE TITLE
Perform system checks during installation

### ISSUE THREAD

**Issue by sizmailov** [textext/textext#54]
Now `setup.py` only copies files into inkscape extensions directory. 

It would be helpful to abort installation if TexText requirements are not met. For example if none of pdflatex/lualatex/xelatex is found in PATH. 

Such checks are expected from `setup.py` https://github.com/textext/textext/issues/51#issuecomment-432431854 This would improve user experience.

Once it's done, installation via `setup.py` should be made the **only** official way of installation of TexText on Linux/Mac OS (and maybe Windows as well?). Abandoning "simple copy" installation method would eliminate fraction of dull error reports.

**Comment by jcwinkler** [comment]
I checked the new setup procedure on Linux. You did a great job @sizmailov , thank you very much!

I will adapt this script to Windows (some paths need to be adapted and the colorama package is required for the color escape sequences). But I will also continue to provide the installer since Windows users are more accustomed to them and you cannot expect a working Python distro on a an average Windows system (TexText uses the Python distro in Inkscape).

Some remarks:
1. I have pstoedit 3.70 and gs 9.25 on my system. However, the output of the script is
```bash
[TexText][SUCCESS ]: |  |  /-and-* pstoedit=3.70 is found
[TexText][ERROR   ]: |  |  /-and-* ghostscript=9.22 is not found (but ghostscript=9.25 is found)
```
indicating an error which in fact is none. Could you fix this? ( in the sense of the logic it is OK of course, but for the user it looks "faulty").

2. Is it possible in the following scenario
```bash
[TexText][SUCCESS ]: |     /--or-* PyGTK2 is found
[TexText][ERROR   ]: |     /--or-* TkInter is not found
[TexText][SUCCESS ]: /-and-+ Detect pdf->png conversion utility
[TexText][SUCCESS ]: |     /--or-* `convert` is found at `/usr/bin`
[TexText][ERROR   ]: |     /--or-* `magick` is NOT found in PATH
```
to mark the "errors" (which are not real errors since they do not affect the functionality of the extension due to the "or") as something which does not appear "faulty". Maybe "Info". Do not know how much effort this takes because it depends on the other operand.

3. If an old installation of TexText is in the extension directory (with the old structure without a textext sub-directory) the old installation remains. I think it would be comfortable to have an option which automatically removes the old "installation".

4. I think the default behavior of the script should be to overwrite an existing installation and to remove an old structure installation without giving command line arguments. A simple question during run-time for overwriting should be more easy to handle by an average user. Maybe a command line option to suppress this question for unattended installation would be nice.

What to you think? Do have time to implement this? I will focus on the windows adaption of the script and the modification of the installer.

Regards, Jan

**Comment by sizmailov** [comment]
1. and 2. Agreed. The best way to address the issue would be traverse requirement-check tree and mark errors as critical or not. The critical errors would be the errors which leads to whole check failure. Non-critical errors can be marked as INFO. Hope I can implement this in a couple of days. 
 3. Good idea. Should be done by default without options. 
 4. Right. Current users already expects `setup.py` to "overwrite" old installation. So we should do it without extra questions/command line options. I think the only concern is ".tex" preamble file. We should try to keep it.

**Comment by sizmailov** [comment]
> I will adapt this script to Windows (some paths need to be adapted and the colorama package is required for the color escape sequences). But I will also continue to provide the installer since Windows users are more accustomed to them and you cannot expect a working Python distro on a an average Windows system (TexText uses the Python distro in Inkscape).

I'm ok with .nsi installation for Windows. By the end of the day `setup.py` might be imperfect and someone would face the need in alternative method. But I think the alternative should be "Plan B". 

Can't we have a wrapper around `setup.py` for windows? If python interpreter comes with inkscape it should be available on any target machine. 

Requirements of `setup.py` itself should be as low as possible. Now it's just a python interpreter (version 2.x or 3.x). Usage of any extra package beyond "builtin" distribution should be optional (need to go through `try import ... except ... `). For example without colorama `setup.py` should print messages in plain text.

**Comment by jcwinkler** [comment]
> If python interpreter comes with inkscape it should be available on any target machine.

The Python interpreter shipped with Inkscape on Windows is not in the system path. Hence calling setup.py from a command window will fail. So on Windows setup.py will be for advanced users and developers. 

> Usage of any extra package beyond "builtin" distribution should be optional (need to go through `try import ... except ... `).

Agreed. Will do so.

**Comment by sizmailov** [comment]
Yes, this Python is not in system path, but it's here. I think we can try, it might be not too elegant what about a .bat file to run setup.py? Should not be a big deal.

**Comment by jcwinkler** [comment]
We could temporarily tweak the system path by the batch file and then run the tests using the Inkscape Python. I will check if the necessary modules are available there. By this approach we can also easily check if the user has installed the Python distro during Inkscape installation (it can be opted out).

**Comment by sizmailov** [comment]
Yes, that is close to what I had in mind. 

> So on Windows setup.py will be for advanced users and developers.

I think new version of `setup.py` brings more value for beginners/people who don't read docs carefully rather than to devs/advanced users. It tells exactly what is wrong with current configuration which is most of time is obvious for latter. 

In addition to error message we should provide a solution. It might be system-specific instructions and/or link to TexText wiki page which address this specific problem.

**Comment by jcwinkler** [comment]
> If an old installation of TexText is in the extension directory (with the old structure without a textext sub-directory) the old installation remains. I think it would be comfortable to have an option which automatically removes the old "installation".

> Right. Current users already expects `setup.py` to "overwrite" old installation. So we should do it without extra questions/command line options. I think the only concern is ".tex" preamble file. We should try to keep it.

I implemented this in the Windows installer. In the case an old installation is detected the old extension files are silently deleted, any tex files are copied to the new dir and the following message box pops up:

![grafik](https://user-images.githubusercontent.com/11866252/47999500-0764d400-e103-11e8-8239-8c68caf4080d.png)

**Comment by sizmailov** [comment]
Looks good! 

>  If you do use TexText as the only LaTeX extension it is safe to click <Yes>

Is there any alternative use case? 

I think preamble is the *only* file we want to keep. The details about copying are redundant. 
In my test question should be shorted to  

"Do you want to keep preamble file from previous installation? Press `<Yes>` to keep, `<No>` to replace"

**Comment by jcwinkler** [comment]
There is the possibility that other extensions have placed tex files into the extension directory as well and it would not be fair to remove them by accident. Also the user could have several package-tex files in the directory so we have to be sure to copy all of them and not only `default_packages.tex`

**Comment by sizmailov** [comment]
I think `default_packages.tex` is our only concern. 
If user created other files, let them be. We should clean up only after previous TexText installations.

I am against doing anything beyond our responsibility. Otherwise we will take risks without knowledge of the situation.

**Comment by leberwurstsaft** [comment]
I think when someone places their .tex files into the extension directory, it's kinda their fault/responsibility. So yes, only take care of the files that came with the old installation.

**Comment by jcwinkler** [comment]
> We could temporarily tweak the system path by the batch file and then run the tests using the Inkscape Python.

I tested this solution and it works in principal. However, colorama package is not available in Inkscape' Python so we have to provide the script without colors or use a different approach, e.g. `curses` library (do not know at the moment if this works toegether with the logger). I will check this and provide a solution.

**Comment by jcwinkler** [comment]
I modified the setup.py so it runs flawlessly on non colored Windows:

https://github.com/jcwinkler/textext/blob/0beb1bea0cfc97c3b96a578a2f4b1bc77baeacbc/setup.py

However, it is not in sync with the newest improvements of @sizmailov. So we need to figure out how to merge it later.

The batch file works in principal (see https://github.com/jcwinkler/textext/blob/0beb1bea0cfc97c3b96a578a2f4b1bc77baeacbc/setup_win.bat), however the following things are missing and have to be carefully considered:
- The path of the Inkscape installation must be properly determined
- Depending on the Inkscape version python.exe is in different subdirs
- Command line options must be passed to setup.py

**Comment by sizmailov** [comment]
I merged changes to https://github.com/sizmailov/textext/tree/merge_setup_py_windows

New: introduced `LinuxDefaults` and `WindowsDefaults` to store all system-dependent values. 

I believe you have a typo in line 33 `"ghostcript": "gs"`, so it's strange that it works on your system unless you have a "ghostscript" executable on Windows otherwise we have a bug in `setup.py`.

Another issue with setup.py on Windows 10. I have installed `ghostscript`, but it's not found by `setup.py`, so it produces false-positive errors. Executable of `ghostscript` named `gswin64c.exe`. When I change `ghostscript` executable name to `gswin64c.exe` all works as expected.  

**Question**: What is correct `ghostscript` executable name on Windows. Is there are many? Is procedure in `win_app_path.py` is better choice? 

Tested replacement of old installation on windows. Feature with replacement of .tex also works. 



> The path of the Inkscape installation must be properly determined
> Depending on the Inkscape version python.exe is in different subdirs

We can try all possible paths one by one in .bat file

> Command line options must be passed to setup.py

Don't think so. The defaults should be good enough.

**Comment by jcwinkler** [comment]
> I believe you have a typo in line 33 `"ghostcript": "gs"`, so it's strange that it works on your system

See my yesterday's commit message of the fix of this typo: https://github.com/jcwinkler/textext/commit/f4f35d914ede393d69295b8fb44211170870168f

But you are right, on Windows it can be `gswin64c.exe` or `gswin32c.exe`. During my rough testing I did not observe this issue because I just checked if the extension is installed which was the case because I have pdf2svg.

> We can try all possible paths one by one in .bat file

There will always be a case where Inkscape has been installed to an unexpected place. I will implement the registry approach, this is not a problem (apart from the crude batch file syntax :smiley: ).

> > Command line options must be passed to setup.py
> 
> Don't think so. The defaults should be good enough.

I disagree. On Windows setup should behave the same as run from Python.

I will open a separate issue which is related to the Windows stuff of the setup.py and will work on it to get all this stuff properly done. So you can focus on the script itself and I will make the Windows part. We can consult each other and later we can crosscheck our results. Is this OK for you?

**Comment by jcwinkler** [comment]
> I merged changes to https://github.com/sizmailov/textext/tree/merge_setup_py_windows
> 
> New: introduced `LinuxDefaults` and `WindowsDefaults` to store all system-dependent values.

Smart solution, thank you!

**Comment by sizmailov** [comment]
Do you mean that `setup.bat` file should forward it's own command line arguments to `setup.py`?
I'm not sure that for regular Windows user it would be easy to run .bat file with command line arguments. 

> I will open a separate issue which is related to the Windows stuff of the setup.py and will work on it to get all this stuff properly done.

Ok, lets do it. In principle we can implement two sets of detection functions as methods `LinuxDefaults` and `WindowsDefaults`.

**Comment by jcwinkler** [comment]
My idea is: An average user who does only want to install the extension should just call `setup.py`, if for special purposes command line arguments are required by advanced users then they can be passed, so basically the same behavior as setup.py has.

**Comment by sizmailov** [comment]
I have implemented almost all features in https://github.com/sizmailov/textext/tree/add_lookup_handles

- [x] Allow users to specify exact paths to executable
  * at setup through command line arguments (e.g: `--xelatex-executable=/usr/bin/xelatex`)
  * after installation by modifying `config.json` (e.g: add  `"xelatex-executable" : "/usr/bin/xelatex"` to `config.json`)
- [x] Create `setup.log` at setup
- [x] Make same checks at setup and runtime. 
- [x] Notify user about bad configuration before any input
- [ ] Provide solutions in case of bad configurations
   - system-dependent fixups (like `To install PyGtk2 run "sudo apt-get install python-gtksoruceview"`)
   - link to wiki for more details

**Comment by sizmailov** [comment]
What should we do with MacOS? It seems to be (the only?) release blocking issue.

In principle it would work, once user provides all paths to executables. But it will never succeed out of box, since `inkscape` is not in PATH by default. A separate MacOS installation page is also required, so all the links will be different.

**Comment by jcwinkler** [comment]
@leberwurstsaft Do you think you may have some time to have a look on this topic on your Mac?

**Comment by leberwurstsaft** [comment]
I was about to suggest that. While I can't help developing at this time, I think I can have a look at Mac compatibility. What exactly should I look for / try?

**Comment by jcwinkler** [comment]
What I learnt from my students is that another crucial part is the detection of the correct Python interpreter, i.e. the interpreter actually called by Inkscape for the execution of the extensions. This is important of course for the installation of PyGTK. If I understood correctly such problems especially arrise when a separate Python distro like Annaconda is installed. On some machines Inkscape called the Anaconda Python and on other machines the system's Python. Had no time for further investigations.

Of course we cannot cover all these variants so we should focus on a precisely defined standard case whatever such a case looks like.

**Comment by sizmailov** [comment]
Inkscape looks for `python` (literally) in PATH (see link below). Most of distributions defines it to be python2.7. For desktop applications PATH variable is not affected by `~/.bashrc`, but affected by `~/.profile`. 

We need to add check for current interpreter version in `__init__.py`. It might be python3 on some distributions. Path to python interpreter might be configured http://wiki.inkscape.org/wiki/index.php/Extension_Interpreters#Selecting_a_specific_interpreter_version_.28via_preferences_file.29

Actually `python-executable` parameter is useful only in `setup.py`. Rest of the time we should utilize actually running interpreter (sys.executable). Can we abandon `--python-executable` parameter? 

I think yes. We need to find a **system** python i.e. from  /usr/bin/ or /bin/

With this solution there is a small chance that user altered either .profile or inkscape settings  and checks at setup.py will be different from runtime. But it would be 
1) negligible fraction of users 
2) who caused the error themselves
3) will be properly instructed at runtime

So I would leave with it. 

I'll implement in a couple of days.

**Comment by sizmailov** [comment]
Update: I think the check should be against currently used python. It's definitely right for TexText runtime. So, lets make it true for setup.py too.

**Comment by jcwinkler** [comment]
I agree. On Windows it is ensured that the correct Python interpreter is called from the batch file.
